sdjwt/src/main/kotlin/web5/security/SdJwt.kt

```diff
@@ -0,0 +1,366 @@
+import com.nimbusds.jwt.JWTClaimsSet
+import com.nimbusds.jwt.SignedJWT
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier
+import java.security.Key
+import java.security.Provider
+import java.security.SignatureException
+
+private const val separator = "~"
+
+internal const val blindedArrayKey = "..."
+
+/**
+ * Represents a Selective Disclosure JWT as defined in https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-05.html#name-terms-and-definitions.
+ * A more detailed overview is available in https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-05.html#name-sd-jwt-structure.
+ *
+ * Note: While [issuerJwt] and [keyBindingJwt] are of type [SignedJWT], they may or may not be signed.
+ */
+public class SdJwt(
+  public val issuerJwt: SignedJWT,
+  public val disclosures: Iterable<Disclosure>,
+  public val keyBindingJwt: SignedJWT? = null,
+  private val serializedSdJwt: String? = null) {
+
+  /** Builder class for constructing a [SdJwt]. */
+  public class Builder(public var issuerHeader: JWSHeader? = null,
+                       public var jwtClaimsSet: JWTClaimsSet? = null,
+                       public var disclosures: Iterable<Disclosure>? = null,
+                       public var keyBindingJwt: SignedJWT? = null) {
+
+    /** Returns an [SdJwt], throwing errors when there are missing values which are required. */
+    public fun build(): SdJwt {
+
+      return SdJwt(
+        SignedJWT(issuerHeader, jwtClaimsSet),
+        disclosures!!,
+        keyBindingJwt,
+      )
+    }
+  }
+
+  /**
+   * Serializes this sd-jwt to the serialization format described in https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-05.html#name-sd-jwt-structure
+   *
+   * [issuerJwt] must have been previously signed.
+   * [keyBindingJwt], if present, must have been previously signed.
+   */
+  @JvmOverloads
+  public fun serialize(mapper: ObjectMapper = defaultMapper): String {
+    if (serializedSdJwt != null) {
+      return serializedSdJwt
+    }
+    require(issuerJwt.state == State.SIGNED) {
+      "issuerJwt is not signed"
+    }
+    if (keyBindingJwt != null) {
+      require(keyBindingJwt.state == State.SIGNED) {
+        "keyBindingJwt is not signed"
+      }
+    }
+    return buildList {
+      add(issuerJwt.serialize())
+      addAll(disclosures.map { it.serialize(mapper) })
+      add(keyBindingJwt?.serialize() ?: "")
+    }.joinToString(separator)
+  }
+
+  /** Signs the [issuerJwt] with [signer]. */
+  @Synchronized
+  public fun signAsIssuer(signer: JWSSigner) {
+    issuerJwt.sign(signer)
+  }
+
+  /**
+   * Signs the [keyBindingJwt] with [signer].
+   */
+  @Synchronized
+  public fun signKeyBinding(signer: JWSSigner) {
+    keyBindingJwt?.sign(signer)
+  }
+
+  /**
+   * TODO: only accept a subset of algos for verification.
+   * Verifies this SD-JWT according to [verificationOptions].
+   */
+  @Throws(Exception::class)
+  public fun verify(
```

Comment:
Maybe add some more detailed explanations on these comments, similar to our other packages, and add example code if it is short and sweet

Reply:
Added more documentation, and an example.